### PR TITLE
chore: Gas optimize

### PR DESCRIPTION
**Pull request by dgusakov**

### CODE REVIEW

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: src/CSModule.sol [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { PausableUntil } from "./lib/utils/PausableUntil.sol";
 import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
 import { AccessControlEnumerableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
 import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import { Arrays } from "@openzeppelin/contracts/utils/Arrays.sol";
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/CSModule.sol]: Fixed

**Review by skhomuti — DISMISSED** [review]

File: src/CSAccounting.sol [hunk 2]
```diff
@@ -574,26 +577,30 @@ contract CSAccounting is
...
-            ) +
-            CSBondLock.getActualLockedBond(nodeOperatorId);
+        unchecked {
+            current = CSBondCore.getBond(nodeOperatorId);
+            required =
+                CSBondCurve.getBondAmountByKeysCount(
+                    nonWithdrawnKeys,
+                    CSBondCurve.getBondCurve(nodeOperatorId)
+                ) +
+                CSBondLock.getActualLockedBond(nodeOperatorId);
```
Comment (madlabman) [inline src/CSAccounting.sol:587]: Add comment that `getActualLockedBond` is uint128, so no overflow expected in practice.

File: src/CSAccounting.sol [hunk 3]
```diff
@@ -607,24 +614,28 @@ contract CSAccounting is
...
-            return nonWithdrawnKeys - bondedKeys;
+            BondCurve memory bondCurve = CSBondCurve.getBondCurve(
+                nodeOperatorId
+            );
+            uint256 bondedKeys = CSBondCurve.getKeysCountByBondAmount(
+                currentBond,
+                bondCurve
+            );
+            return
+                bondedKeys < nonWithdrawnKeys
```
Comment (madlabman) [inline src/CSAccounting.sol:636]:
```suggestion
                nonWithdrawnKeys > bondedKeys
```

File: src/CSModule.sol [hunk 4]
```diff
@@ -944,11 +944,12 @@ contract CSModule is
         uint256 amount
     ) external onlyRole(REPORT_EL_REWARDS_STEALING_PENALTY_ROLE) {
         _onlyExistingNodeOperator(nodeOperatorId);
-        accounting.lockBondETH(
-            nodeOperatorId,
-            amount + EL_REWARDS_STEALING_FINE
-        );
-
+        unchecked {
```
Comment (madlabman) [inline src/CSModule.sol:947]: Remove here because of a possible overflow.

File: src/CSAccounting.sol [hunk 5]
```diff
@@ -477,37 +475,41 @@ contract CSAccounting is
...
+            uint256 requiredForNextKeys = CSBondCurve.getBondAmountByKeysCount(
+                nonWithdrawnKeys + additionalKeys,
+                bondCurve
+            ) -
+                CSBondCurve.getBondAmountByKeysCount(
+                    nonWithdrawnKeys,
+                    bondCurve
+                );
+
+            uint256 missing = required > current ? required - current : 0;
```
Comment (madlabman) [inline src/CSAccounting.sol:498]: Move `unchecked` here.

File: src/CSModule.sol [hunk 6]
```diff
@@ -990,9 +991,10 @@ contract CSModule is
     /// @dev SETTLE_EL_REWARDS_STEALING_PENALTY_ROLE role is expected to be assigned to Easy Track
     /// @param nodeOperatorIds IDs of the Node Operators
     function settleELRewardsStealingPenalty(
-        uint256[] memory nodeOperatorIds
+        uint256[] calldata nodeOperatorIds
     ) external onlyRole(SETTLE_EL_REWARDS_STEALING_PENALTY_ROLE) {
-        for (uint256 i; i < nodeOperatorIds.length; ++i) {
+        uint256 noIdCount = nodeOperatorIds.length;
```
Comment (madlabman) [inline src/CSModule.sol:996]: Double check gas usage.

File: src/CSModule.sol [hunk 7]
```diff
@@ -1066,7 +1068,9 @@ contract CSModule is
         }
 
         if (amount < DEPOSIT_SIZE) {
```
Comment (madlabman) [inline src/CSModule.sol:1070]:
```suggestion
        if (DEPOSIT_SIZE > amount) {
```

File: src/CSModule.sol [hunk 8]
```diff
@@ -1473,14 +1482,15 @@ contract CSModule is
             ? limit
             : nodeOperatorsCount - offset;
         nodeOperatorIds = new uint256[](idsCount);
-        for (uint256 i = 0; i < nodeOperatorIds.length; ++i) {
+        uint256 len = nodeOperatorIds.length;
```
Comment (madlabman) [inline src/CSModule.sol:1485]: Double check please.

File: src/CSModule.sol [hunk 9]
```diff
@@ -1575,9 +1584,10 @@ contract CSModule is
...
             revert SigningKeysInvalidOffset();
         }
-
-        if (startIndex + keysCount > no.totalAddedKeys) {
-            revert SigningKeysInvalidOffset();
+        unchecked {
+            if (startIndex + keysCount > no.totalAddedKeys) {
+                revert SigningKeysInvalidOffset();
+            }
         }
```
Comment (madlabman) [inline src/CSModule.sol:1591]: Remove the check.

Comment (madlabman) [reply src/CSModule.sol:1591]: Remove `unchecked` in `removeKeysSigs` or double check.

File: src/lib/SigningKeys.sol [hunk 10]
```diff
@@ -91,65 +94,70 @@ library SigningKeys {
...
-        for (uint256 i = startIndex + keysCount; i > startIndex; ) {
-            curOffset = SIGNING_KEYS_POSITION.getKeyOffset(
-                nodeOperatorId,
-                i - 1
-            );
-            assembly {
-                // read key
-                mstore(add(tmpKey, 0x30), shr(128, sload(add(curOffset, 1)))) // bytes 16..47
-                mstore(add(tmpKey, 0x20), sload(curOffset)) // bytes 0..31
+        unchecked {
```
Comment (madlabman) [inline src/lib/SigningKeys.sol:104]: Maybe to remove.

File: src/CSVerifier.sol [hunk 11]
```diff
@@ -20,7 +20,9 @@ function amountWei(Withdrawal memory withdrawal) pure returns (uint256) {
 /// @notice Convert gwei to wei
 /// @param amount Amount in gwei
 function gweiToWei(uint64 amount) pure returns (uint256) {
-    return uint256(amount) * 1 gwei;
+    unchecked {
+        return uint256(amount) * 1 gwei;
```
Comment (madlabman) [inline src/CSVerifier.sol:24]: Add comment.

File: src/abstract/CSBondCurve.sol [hunk 12]
```diff
@@ -137,17 +142,19 @@ abstract contract CSBondCurve is ICSBondCurve, Initializable {
         BondCurve memory curve
     ) public pure returns (uint256) {
         if (amount < curve.points[0]) return 0;
```
Comment (madlabman) [inline src/abstract/CSBondCurve.sol:144]: unafeMemoryAccess ?

File: src/lib/SigningKeys.sol [hunk 13]
```diff
@@ -27,24 +27,27 @@ library SigningKeys {
...
     ) internal returns (uint256) {
-        if (keysCount == 0 || startIndex + keysCount > UINT64_MAX) {
-            revert InvalidKeysCount();
-        }
-        if (
-            pubkeys.length != keysCount * PUBKEY_LENGTH ||
-            signatures.length != keysCount * SIGNATURE_LENGTH
-        ) {
-            revert InvalidLength();
+        unchecked {
```
Comment (madlabman) [inline src/lib/SigningKeys.sol:41]: Check again.

**Review by madlabman — CHANGES_REQUESTED** [review]

File: src/CSModule.sol (hunk 6 above)
Comment (dgusakov) [reply src/CSModule.sol:996]: Without this change, it is even cheaper. lol

File: src/lib/SigningKeys.sol (hunk 10 above)
Comment (dgusakov) [reply src/lib/SigningKeys.sol]: reconsidered and moved

File: src/lib/SigningKeys.sol (hunk 13 above)
Comment (dgusakov) [reply src/lib/SigningKeys.sol:44]: reconsidered and moved

File: src/abstract/CSBondLock.sol [hunk 14]
```diff
@@ -59,6 +62,10 @@ abstract contract CSBondLock is ICSBondLock, Initializable {
         if (minBondLockRetentionPeriod > maxBondLockRetentionPeriod) {
             revert InvalidBondLockRetentionPeriod();
         }
+        // retention period can not be more than type(uint64).max to avoid overflow when setting bond lock
+        if (maxBondLockRetentionPeriod > type(uint64).max) {
```
Comment (madlabman) [inline src/abstract/CSBondLock.sol:66]: uint128?

Comment (dgusakov) [reply src/abstract/CSBondLock.sol:66]: Within the BondLock structure, we have a field called retentionUntil. The value for this field is calculated as `currentTimestamp + retentionPeriod`, hence limiting to uint128 here will not be sufficient

**Review by madlabman — APPROVED** [review]

File: src/CSModule.sol [hunk 15]
```diff
@@ -1531,12 +1534,15 @@ contract CSModule is
...
         NodeOperator storage no = _nodeOperators[nodeOperatorId];
         uint256 startIndex = no.totalAddedKeys;
-        if (
-            !publicRelease &&
-            startIndex + keysCount >
-            MAX_SIGNING_KEYS_PER_OPERATOR_BEFORE_PUBLIC_RELEASE
-        ) {
-            revert MaxSigningKeysCountExceeded();
+        unchecked {
+            // startIndex + keysCount can overflow here but will revert later in the SigningKeys.saveKeysSigs
```
Comment (skhomuti) [inline src/CSModule.sol:1538]: cannot because of deposit check before

**Review by skhomuti — APPROVED** [review]

**Review by skhomuti — APPROVED** [review]

**Review by vgorkavenko — APPROVED** [review]

**Review by madlabman — APPROVED** [review]

**Pull request merged**